### ISSUE TITLE
Pylint C0103 (invalid-name) contradiction for class-level typing.Final attributes.

### ISSUE THREAD

**Issue by xiahualiu-txg** [pylint-dev/pylint#10711]
### Bug description

Pylint (C0103) generates contradictory `invalid-name` warnings for class-level attributes annotated with `typing.Final`.

* It flags a `snake_case` attribute annotated with `Final` as an invalid constant (demanding `UPPER_CASE`).
* It flags an `UPPER_CASE` attribute annotated with `Final` as an invalid attribute (demanding `snake_case`).

This creates a catch-22 where no naming convention is considered valid.

## A Simple Example

```python
# file: example.py
from dataclasses import dataclass
from typing import Final

@dataclass
class ExampleClass:
    parameter_one: int = 10
    parameter_two: str = "default"
    parameter_three: float = 3.14

    # C0103: Class constant name "parameter_six" doesn't conform to UPPER_CASE naming style (invalid-name)
    parameter_six: Final[int] = 42

    # C0103: Attribute name "PARAMETER_FIVE" doesn't conform to snake_case naming style (invalid-name)
    PARAMETER_FIVE: Final[int] = 42
```

## Analysis

The core purpose of both the `UPPER_CASE` naming convention (from PEP 8) and the `typing.Final` annotation is identical: to signal developer intent that a variable is a constant and should not be reassigned.

Pylint's linter appears to be applying two different rules inconsistently:

For `parameter_six`, it correctly identifies `Final` as indicating a constant and therefore demands the `UPPER_CASE` naming convention.

For `PARAMETER_FIVE`, it seems to ignore the `Final` annotation, classify the variable as a standard attribute (which can be modified), and therefore incorrectly demands the `snake_case` naming convention.

This logic is contradictory. The `UPPER_CASE` convention was created precisely to communicate the intent of immutability that `typing.Final` now enforces statically. These two signals should be seen as compatible and reinforcing, not mutually exclusive.

## Proposed Behavior

When a class-level attribute is annotated with `typing.Final`, it should be treated as a constant for naming convention purposes.

* `PARAMETER_FIVE: Final[int] = 42` (UPPER_CASE) should NOT raise an invalid-name warning. It correctly uses the constant naming convention.
* `parameter_six: Final[int] = 42` (snake_case) SHOULD (as it currently does) raise an invalid-name warning, suggesting the `UPPER_CASE` convention is preferred for `Final` attributes.

I suggest Pylint treats `UPPER_CASE` same as `Final` because they are basically flagging the same thing in the code, that is DO NOT modify this. The logic behind because `UPPER_CASE` attributes can be modified so it needs to be `snake_case` reverses the logic.

### Configuration

```ini
pylint 3.3.8
python 3.11.14
```

### Command used

```shell
python -m pylint /home/ubuntu/GitHub/embedded-apps/example.py
```

### Pylint output

```python
************* Module example
example.py:21:4: C0103: Attribute name "PARAMETER_FIVE" doesn't conform to snake_case naming style (invalid-name)
example.py:18:4: C0103: Class constant name "parameter_six" doesn't conform to UPPER_CASE naming style (invalid-name)
```

### Expected behavior

The first "PARAMETER_FIVE" is legit `UPPER_CASE` user case so pylint shouldn't raise anything.

### Pylint version

```shell
pylint 3.3.8
```

### OS / Environment

Ubuntu 24.04.3

### Additional dependencies

```python
No.
```

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening an issue. This behavior recentely changed in pylint 4.0, and we don't support pylint 3 anymore, can you still reproduce with pylint 4+ ?

**Comment by xiahualiu-txg** [comment]
It can be reproduced on pylint 4.0.2, I just tested using the same example, the result is the same:

```
(.venv) ubuntu@1322938a9a15:~/GitHub$ python -m pylint ./example.py                     
************* Module example
example.py:18:4: C0103: Attribute name "PARAMETER_FIVE" doesn't conform to snake_case naming style (invalid-name)
example.py:15:4: C0103: Class constant name "parameter_six" doesn't conform to UPPER_CASE naming style (invalid-name)

------------------------------------------------------------------
Your code has been rated at 7.78/10 (previous run: 8.00/10, -0.22)

(.venv) ubuntu@1322938a9a15:~/GitHub$ python -m pylint --version    
pylint 4.0.2
astroid 4.0.1
Python 3.11.14 (main, Oct 10 2025, 08:54:04) [GCC 13.3.0]
```

**Comment by Pierre-Sassoulas** [comment]
Indeed, this is reproducible on main.

**Comment by xiahualiu-txg** [comment]
@Pierre-Sassoulas Sorry to bother you again on this.

A day after posting this issue, I realized this is another issue:

```
# C0103: Class constant name "parameter_six" doesn't conform to UPPER_CASE naming style (invalid-name)
parameter_six: Final[int] = 42
```

The issue is that Pylint currently flags non-`UPPER_CASE` names as `invalid-name` when they are explicitly marked with immutability qualifiers like `Final`.

I suggest Pylint **NOT** require the `UPPER_CASE` naming style when an attribute or variable is decorated with an type qualifier that signals constancy or immutability, such as:

* `typing.Final`
* `typing.ClassVar` (when used on a class attribute)
* `typing.ReadOnly` (e.g., in `TypedDict` or as a future attribute qualifier)
* `dataclasses.dataclass(frozen=True)` (for all attributes within the class)

## Reason 1

`UPPER_CASE` predates the type system and has become an established PEP 8 idiom for a value intended to be a **static constant**. This convention implies a stronger, more restrictive meaning than what the current type annotations enforce.

`UPPER_CASE` should signify a value that is set in the static code and should **never be modified at runtime**. This intent goes beyond `ClassVar` (which is mutable) or `Final`, `frozen=True` (which can be set in `__init__`). Type checkers should implicitly enforce `Final` or `ReadOnly` when an attribute is named `UPPER_CASE`, but not vice versa.

* `UPPER_CASE`: **static constant** that is set in code and should never be other values during runtime, but it doesn't enforce `ClassVar`, you can still have per-instance static constants for better cache locality.
* `Final`, `frozen=True`: **immuntable attributes** that can be set per-instance or per-classtype.

## Resaon 2 

If `UPPER_CASE` needs type annotations, then:

1. It will be very hard to make a decision which type hints are required for `UPPER_CASE` names, do I need `frozen=True`, `ClassVar` or `Final` or do I need all of them to justify that I need to use `UPPER_CASE` or I can use `UPPER_CASE` names?
2. For different classes `TypedDict`, `dataclass`, etc. different immuntable type hints like `ReadOnly` are used there, making it even harder to make thing right.

So that's why I suggest that maybe `pylint` can seperate `UPPER_CASE` from these type hints so one doesn't rely on another? I am open for different opinions here although, above is just my random thought after I created this ticket.

**Comment by Pierre-Sassoulas** [comment]
You can make suggestion to add test case and give your opinion for existing one on the functional test case in #10712
